File: _Src/Freem/Freem.Entities.UseCases.Contracts/Records/Create/CreateRecordRequest.cs
using Freem.Entities.Records.Models;
using Freem.Entities.Relations.Collections;
using Freem.Time.Models;

namespace Freem.Entities.UseCases.Contracts.Records.Create;

public sealed class CreateRecordRequest
{
    private readonly RelatedTagsCollection _tags = RelatedTagsCollection.Empty;
    
    public DateTimePeriod Period { get; }

    public RecordName? Name { get; init; }
    public RecordDescription? Description { get; init; }
    
    public RelatedActivitiesCollection Activities { get; }

    public RelatedTagsCollection Tags
    {
        get => _tags;
        init
        {
            ArgumentNullException.ThrowIfNull(value);
            
            _tags = value;
        }
    } 
    
    public CreateRecordRequest(DateTimePeriod period, RelatedActivitiesCollection activities)
    {
        ArgumentNullException.ThrowIfNull(period);
        ArgumentNullException.ThrowIfNull(activities);
        
        Period = period;
        Activities = activities;
    }
}

File: _Src/Freem/Freem.Web/src/clients/models/records/CreateRecordRequest.ts
export class CreateRecordRequest {
  public readonly period: string;
  public readonly activities: string[];

  public readonly name?: string;
  public readonly description?: string;
  public readonly tags?: string[];

  constructor(
    period: string, 
    activities: string[],
    name?: string, 
    description?: string, 
    tags?: string[]
  ) {
    this.period = period;
    this.activities = activities;
    
    this.name = name;
    this.description = description;
    this.tags = tags;
  }
}